ul/src/pdu/writer.rs

```diff
@@ -0,0 +1,745 @@
+use crate::error::Result;
+use crate::pdu::*;
+use byteordered::byteorder::{BigEndian, WriteBytesExt};
+use std::io::Write;
+
+fn write_n(writer: &mut dyn Write, bytes_to_write: &[u8]) -> Result<()> {
+    let mut pos = 0;
+    while pos < bytes_to_write.len() {
+        let bytes_written = writer.write(&bytes_to_write[pos..])?;
+        if bytes_written == 0 {
+            return Err(std::io::Error::new(
+                std::io::ErrorKind::WriteZero,
+                "wrote zero length",
+            ))?;
+        }
+        pos += bytes_written;
+    }
+    Ok(())
+}
+
+pub(crate) fn write_chunk_u32<F>(writer: &mut dyn Write, func: F) -> Result<()>
```

Comment:
`func` is always called with a `&mut Vec<u8>`. If you make that transparent into `F`, I guess that you will no longer need trait objects.